src/replay/service.zig

```diff
@@ -31,151 +36,492 @@ const MAX_ENTRIES: u64 = 1024 * 1024; // 1 million slots is about 5 days
+    logger: Logger,
+    my_identity: Pubkey,
+    vote_identity: Pubkey,
     /// Tell replay when to exit
     exit: *std.atomic.Value(bool),
     /// Used in the EpochManager
     epoch_schedule: sig.core.EpochSchedule,
+    account_store: sig.accounts_db.AccountStore,
+    /// Reader used to get the entries to validate them and execute the transactions
+    /// Writer used to update the ledger with consensus results
+    ledger: LedgerRef,
     /// Used to get the entries to validate them and execute the transactions
-    blockstore_reader: *BlockstoreReader,
-    /// Used to update the ledger with consensus results
-    ledger_result_writer: *LedgerResultWriter,
-    account_store: AccountStore,
     slot_leaders: SlotLeaders,
+    senders: Senders,
+    receivers: Receivers,
+
     /// The slot to start replaying from.
-    root_slot: Slot,
-    root_slot_constants: sig.core.SlotConstants,
-    root_slot_state: sig.core.SlotState,
+    root: struct {
+        slot: Slot,
+        constants: sig.core.SlotConstants,
+        state: sig.core.SlotState,
+    },
 };
 
-pub const Logger = sig.trace.ScopedLogger("replay");
+pub const LedgerRef = struct {
+    db: BlockstoreDB,
+    reader: *sig.ledger.BlockstoreReader,
+    writer: *sig.ledger.LedgerResultWriter,
+};
+
+pub const SlotHistoryAccessor = struct {
+    account_reader: sig.accounts_db.AccountReader,
+
+    pub fn init(account_reader: sig.accounts_db.AccountReader) SlotHistoryAccessor {
+        return .{
+            .account_reader = account_reader,
+        };
+    }
 
-pub const SlotData = struct {
-    duplicate_confirmed_slots: replay.edge_cases.DuplicateConfirmedSlots,
-    epoch_slots_frozen_slots: replay.edge_cases.EpochSlotsFrozenSlots,
-    duplicate_slots_to_repair: replay.edge_cases.DuplicateSlotsToRepair,
-    purge_repair_slot_counter: replay.edge_cases.PurgeRepairSlotCounters,
-    unfrozen_gossip_verified_vote_hashes: replay.edge_cases.UnfrozenGossipVerifiedVoteHashes,
-    duplicate_slots: replay.edge_cases.DuplicateSlots,
+    pub fn getSlotHistory(
+        self: SlotHistoryAccessor,
+        allocator: std.mem.Allocator,
+        ancestors: *const Ancestors,
+    ) !SlotHistory {
+        const slot_history: SlotHistory = sh: {
+            const maybe_account = try self.account_reader.forSlot(ancestors).get(SlotHistory.ID);
+            const account: sig.core.Account = maybe_account orelse
+                return error.MissingSlotHistoryAccount;
+
+            var data_iter = account.data.iterator();
+            const slot_history = try sig.bincode.read(
+                allocator,
+                SlotHistory,
+                data_iter.reader(),
+                .{},
+            );
+            errdefer slot_history.deinit(allocator);
+
+            if (data_iter.bytesRemaining() != 0) {
+                return error.TrailingBytesInSlotHistory;
+            }
+
+            break :sh slot_history;
+        };
+        return slot_history;
```

Comment:
You can flatten this block to be the same as the function body with only a return and no break.